lib/Sema/DerivedConformanceEquatableHashable.cpp

```diff
@@ -1104,33 +1139,85 @@ bool DerivedConformance::canDeriveHashable(TypeChecker &tc,
+        return fd;
+    }
+  }
+  return nullptr;
+}
+
+static ProtocolConformance *
+getHashableConformance(Decl *parentDecl) {
+  ASTContext &C = parentDecl->getASTContext();
+  auto DC = cast<DeclContext>(parentDecl);
+  auto hashableProto = C.getProtocol(KnownProtocolKind::Hashable);
+  for (auto conformance: DC->getLocalConformances()) {
+    if (conformance->getProtocol() == hashableProto) {
+      return conformance;
+    }
+  }
+  return nullptr;
+}
+
 ValueDecl *DerivedConformance::deriveHashable(TypeChecker &tc,
                                               Decl *parentDecl,
                                               NominalTypeDecl *type,
                                               ValueDecl *requirement) {
+  ASTContext &C = parentDecl->getASTContext();
+  auto hashableProto = C.getProtocol(KnownProtocolKind::Hashable);
+
   // Conformance can't be synthesized in an extension; we allow it as a special
   // case for enums with no associated values to preserve source compatibility.
   auto theEnum = dyn_cast<EnumDecl>(type);
   if (!(theEnum && theEnum->hasOnlyCasesWithoutAssociatedValues()) &&
       type != parentDecl) {
-    auto hashableProto = tc.Context.getProtocol(KnownProtocolKind::Hashable);
     auto hashableType = hashableProto->getDeclaredType();
     tc.diagnose(parentDecl->getLoc(), diag::cannot_synthesize_in_extension,
                 hashableType);
     return nullptr;
   }
 
-  // Build the necessary decl.
-  if (requirement->getBaseName() == "hashValue") {
-    if (theEnum)
-      return deriveHashable_hashValue(tc, parentDecl, theEnum,
-                                      &deriveBodyHashable_enum_hashValue);
-    else if (auto theStruct = dyn_cast<StructDecl>(type))
-      return deriveHashable_hashValue(tc, parentDecl, theStruct,
-                                      &deriveBodyHashable_struct_hashValue);
-    else
-      llvm_unreachable("todo");
+  // var hashValue: Int
+  if (requirement->getBaseName() == C.Id_hashValue) {
+    auto hashValueDecl = deriveHashable_hashValue(tc, parentDecl, type);
+
+    // Also derive _hash(into:) -- it has a default implementation, so we
+    // wouldn't otherwise consider it as a candidate for synthesizing.
+    //
+    // FIXME: This assumes that _hash(into:) hasn't already been resolved. It
+    // would be nicer to remove the default implementation and independently
+    // synthesize either/both of the two Hashable requirements as needed;
+    // however, synthesizing methods (as opposed to properties) into imported
+    // types is not yet fully supported.
+    if (auto hashIntoReq = getHashIntoRequirement(C)) {
+      AbstractFunctionDecl::BodySynthesizer hashIntoSynthesizer;
+      if (theEnum)
+        hashIntoSynthesizer = &deriveBodyHashable_enum_hashInto;
+      else if (auto theStruct = dyn_cast<StructDecl>(type))
+        hashIntoSynthesizer = &deriveBodyHashable_struct_hashInto;
+      else
+        llvm_unreachable("");
```

Comment:
Please leave an actual comment here, like "attempt to derived hashInto for a type other than a struct or enum".